cpp/src/Slice/MetadataValidator.cpp

```diff
@@ -0,0 +1,410 @@
+    {
+        // Special case: if this metadata can _only_ be applied to types, but it was applied to an operation,
+        // what this really means is that the metadata applies to that operation's return type.
+        // Same thing if this metadata has been applied to a parameter or data member as well.
+        if (auto op = dynamic_pointer_cast<Operation>(p); op && info.isTypeMetadata)
+        {
+            if (const auto returnType = op->returnType())
+            {
+                appliedTo = returnType;
+            }
+            else
+            {
+                string msg = '\'' + directive + "' metadata cannot be applied operations with void return type";
+                p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
+                isValid = false;
+            }
+        }
+        else if (auto param = dynamic_pointer_cast<ParamDecl>(p); param && info.isTypeMetadata)
+        {
+            appliedTo = param->type();
+        }
+        else if (auto dm = dynamic_pointer_cast<DataMember>(p); dm && info.isTypeMetadata)
+        {
+            appliedTo = dm->type();
+        }
+        // Otherwise this metadata is just invalid.
+        else
+        {
+            string msg = '\'' + directive + "' metadata cannot be applied to definitions and declarations" +
+                         ", it can only be applied directly to types";
+            p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
+            isValid = false;
+        }
+    }
+    else
+    {
+        // Otherwise if there's nothing special going on, then the metadata was applied directly to `p`.
+        appliedTo = p;
+    }
+
+    if (appliedTo)
+    {
+        const list<const type_info*>& validOn = info.validOn;
+        auto found =
+            std::find_if(validOn.begin(), validOn.end(), [&](const type_info* t) { return *t == typeid(*appliedTo); });
+        if (!validOn.empty() && found == validOn.end())
+        {
+            string message = misappliedMetadataMessage(metadata, appliedTo);
+            p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, message);
+            isValid = false;
+        }
+    }
+
+    // Fifth, we check if this metadata is a duplicate, ie. has the same directive already been applied in this context?
+    if (info.mustBeUnique)
+    {
+        // 'insert' only returns `true` if the value wasn't already present in the set.
+        bool wasInserted = _seenDirectives.insert(directive).second;
+        if (!wasInserted)
+        {
+            string msg = "ignoring duplicate metadata: '" + directive + "' has already been applied in this context";
+            p->unit()->warning(metadata->file(), metadata->line(), InvalidMetadata, msg);
+            isValid = false;
+        }
+    }
+
+    // Finally, if a custom validation function is specified for this metadata, we run it.
+    if (info.extraValidation.has_value())
```

Comment:
For a nullable types like `std::function` I'lll just use ` nullptr` as not set instead of an optional?